## 一、Python语法

#### 1.九九乘法表

In [1]:
# ①
for i in range(1,10):
    for j in range(1,i+1):
        print('{}x{}={}\t'.format(i,j,i*j),end='')
    print()

1x1=1	
2x1=2	2x2=4	
3x1=3	3x2=6	3x3=9	
4x1=4	4x2=8	4x3=12	4x4=16	
5x1=5	5x2=10	5x3=15	5x4=20	5x5=25	
6x1=6	6x2=12	6x3=18	6x4=24	6x5=30	6x6=36	
7x1=7	7x2=14	7x3=21	7x4=28	7x5=35	7x6=42	7x7=49	
8x1=8	8x2=16	8x3=24	8x4=32	8x5=40	8x6=48	8x7=56	8x8=64	
9x1=9	9x2=18	9x3=27	9x4=36	9x5=45	9x6=54	9x7=63	9x8=72	9x9=81	


In [3]:
# ②
for i in range(1,10):
    for j in range(1,i+1):
        print("%dx%d=%d" % (i,j,i*j),end=" ")
    print (" ")

1x1=1  
2x1=2 2x2=4  
3x1=3 3x2=6 3x3=9  
4x1=4 4x2=8 4x3=12 4x4=16  
5x1=5 5x2=10 5x3=15 5x4=20 5x5=25  
6x1=6 6x2=12 6x3=18 6x4=24 6x5=30 6x6=36  
7x1=7 7x2=14 7x3=21 7x4=28 7x5=35 7x6=42 7x7=49  
8x1=8 8x2=16 8x3=24 8x4=32 8x5=40 8x6=48 8x7=56 8x8=64  
9x1=9 9x2=18 9x3=27 9x4=36 9x5=45 9x6=54 9x7=63 9x8=72 9x9=81  


#### 2.List操作

In [15]:
# 参考http://www.runoob.com/python/python-exceptions.html
def square_list(a_list):
    try:
        return sum([i**2 for i in a_list])
    except:
        print("请输入数字")
square_list([1,2,3,4]) 

30

In [14]:
a_list=eval(input("请输入list:"))
square_list(a_list)

请输入list:['a',1,2]
请输入数字


#### 3.文本IO操作

In [67]:
# 参考http://www.runoob.com/python/python-files-io.html
# 写一个data.txt
with open("data.txt","w") as f:
    num_total=eval(input("你想输入多少个整数："))
    for i in range(num_total):
        int_num=input("输入整数：")
        f.write(int_num+"\n")

你想输入多少个整数：6
输入整数：8
输入整数：9
输入整数：5
输入整数：4
输入整数：10
输入整数：5


In [73]:
# 读data.txt，写入data_asc.txt
number=[]
for line in open("data.txt","r"):
    line = eval(line.rstrip('\n'))
    number.append(line)
asc_num=sorted(number)
print(asc_num)
with open("data_asc.txt","w") as f:
    for i in asc_num:
        f.write(str(i)+"\n")       

[4, 5, 5, 8, 9, 10]


In [74]:
# 查看一下
with open("data_asc.txt","r") as f:
    something=f.read()
    print(something)

4
5
5
8
9
10



#### 4.数字操作

In [77]:
# ①
for i in range(1,9):
    for j in range(2,10):
        if i<j :
            print('{}{}\t'.format(i,j),end='')
    print()

12	13	14	15	16	17	18	19	
23	24	25	26	27	28	29	
34	35	36	37	38	39	
45	46	47	48	49	
56	57	58	59	
67	68	69	
78	79	
89	


In [84]:
# ②
for num in range(10,100):
    if (num/10)<(num%10):
        print('{}\t'.format(num),end='')

12	13	14	15	16	17	18	19	23	24	25	26	27	28	29	34	35	36	37	38	39	45	46	47	48	49	56	57	58	59	67	68	69	78	79	89	

## 二、观看吴恩达ML视频

## 三、感想与看法

## 四、Numpy 和 Pandas

## 五、读取xlsx文件操作

In [93]:
# 参考https://blog.csdn.net/index20001/article/details/73844576
import xlrd
 
workbook = xlrd.open_workbook('data.xlsx')
booksheet = workbook.sheet_by_index(0)         #用索引取第一个sheet

#读取第二行数据
row_2 = booksheet.row_values(1)
print("读取第二行：",row_2)
#读取第一列数据
column_1=booksheet.col_values(0)
print("读取第一列：",column_1)
#读第一行第六列单元格数据
cell_1_6 = booksheet.cell_value(0,5)
print("读取第一行第六个字符：",cell_1_6)

读取第二行： [0.0, 1.2, 200.0, 6.0, '', 0.0]
读取第一列： ['X1', 0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0]
读取第一行第六个字符： Y


## 六、PPT展示成果

## 选做：线性拟合
选用的数据为scikit-learn自带的波士顿房价数据集

#### 1.首先，先导入数据

In [94]:
from sklearn.datasets import load_boston

boston=load_boston()
X=boston.data
y=boston.target
X.shape     # 输出数据集的大小

(506, 13)

In [97]:
X[0]       # 查看一下样本数据

array([6.320e-03, 1.800e+01, 2.310e+00, 0.000e+00, 5.380e-01, 6.575e+00,
       6.520e+01, 4.090e+00, 1.000e+00, 2.960e+02, 1.530e+01, 3.969e+02,
       4.980e+00])

In [98]:
boston.feature_names    # 查看一下特征的标签

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

#### 2.模型训练
将数据的20%为测试集，80%为训练集

In [100]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=3)

In [103]:
import time
from sklearn.linear_model import LinearRegression

model=LinearRegression()
start=time.clock()             # 用于计算时间
model.fit(X_train,y_train)     # 模型训练

train_score=model.score(X_train,y_train)    # 模型对训练样本的准确率得分
cv_score=model.score(X_test,y_test)         # 模型对测试样本的准确率得分
print('elaspe:{0:.6f};train_score:{1:0.6f};cv_score:{2:.6f}'.format(time.clock()-start,train_score,cv_score))

elaspe:0.004071;train_score:0.723941;cv_score:0.794958


#### 结果分析：
    从得分情况来看，模型拟合效果一般！训练样本的评分较低（train_score：0.723941），猜测欠拟合，考虑增大多项式阶数来优化模型

#### 3.模型优化

In [104]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline

def polynomial_model(degree=1):
    polynomial_features=PolynomialFeatures(degree=degree,include_bias=False)
    linear_regression=LinearRegression(normalize=True)
    pipeline=Pipeline([("polynomial_features",polynomial_features),("linear_regression",linear_regression)])
    return pipeline

用二阶多项式来拟合

In [105]:
model=polynomial_model(degree=2)

start=time.clock()             # 用于计算时间
model.fit(X_train,y_train)     # 模型训练
train_score=model.score(X_train,y_train)    # 模型对训练样本的准确率得分
cv_score=model.score(X_test,y_test)         # 模型对测试样本的准确率得分
print('elaspe:{0:.6f};train_score:{1:0.6f};cv_score:{2:.6f}'.format(time.clock()-start,train_score,cv_score))

elaspe:0.308509;train_score:0.930547;cv_score:0.860465


用三阶多项式来拟合

In [106]:
model=polynomial_model(degree=3)

start=time.clock()             # 用于计算时间
model.fit(X_train,y_train)     # 模型训练
train_score=model.score(X_train,y_train)    # 模型对训练样本的准确率得分
cv_score=model.score(X_test,y_test)         # 模型对测试样本的准确率得分
print('elaspe:{0:.6f};train_score:{1:0.6f};cv_score:{2:.6f}'.format(time.clock()-start,train_score,cv_score))

elaspe:0.408610;train_score:1.000000;cv_score:-101.629888


#### 结果分析
    二阶多项式模型拟合效果有明显提高！
    三阶多项式的测试样本得分为负数，出现过拟合！
    由此：二阶多项式模型较为合适